### PR TITLE
Add `ArchivematicaAPI` and beginnings of `run_dark_blue.rb` (DBAI-3)

### PR DESCRIPTION
**Pull request by ssciolla**
This PR aims to resolve [DBAI-3](https://mlit.atlassian.net/jira/software/projects/DBAI/boards/89?selectedIssue=DBAI-3).

**Testing notes**

Run the unit tests for `ArchivematicaAPI` with
```sh
ruby test/test_archivematica.rb
```

Run the beginnings of the dark blue process with
```sh
ruby run_dark_blue.rb
```
Note that you'll need to be using the U-M VPN or on campus.

**To Do**

- [x] Add basic implementation of `ArchivematicaAPI` class
- [x] Add configuration handling for the Archivematica REST API
- [x] Add `run_dark_blue.rb` with `Archivematica` usage, some class setup, and comments
- [x] Add error handling
- [x] Add retry capability
- [x] Add tests for `ArchivematicaAPI`

### CODE REVIEW

File: test/test_archivematica.rb [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+    ) do |builder|
+      builder.request :retry
+      builder.response :raise_error
+
+      builder.adapter :test, @stubs
+    end
+    @stubbed_api = ArchivematicaAPI.new(
+      stubbed_test_conn,
+      api_prefix: api_prefix
+    )
```
Comment (ssciolla) [inline test/test_archivematica.rb:66]: In the course of writing tests, I learned about the [Faraday test adapter](https://lostisland.github.io/faraday/#/adapters/test-adapter), which seemed suitable for the error handling and retry tests. However, I couldn't figure out how to add the test adapter to an existing connection, which mean that I had to inject the `Connection` in the constructor. But that leads to some awkwardness here where the test setup MAKES the test retry and error handling pass, though it's really just reproducing what would be done regularly with `ArchivematicaAPI.from_config`. Don't know if this is something to worry about or just let it go. Maybe [`httpx`](https://github.com/HoneyryderChuck/httpx) can do this better?

File: lib/archivematica.rb [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+      end
+      results
+    end
+
+    def get_packages(location_uuid:)
+      package_objects = get_objects_from_pages(PACKAGE_PATH, {
+        "current_location" => location_uuid
+      })
+      LOGGER.debug("Total number of package objects found: #{package_objects.length}")
+      packages = package_objects.select { |o| o["status"] == PackageStatus::UPLOADED }.map do |o|
```
Comment (ssciolla) [inline lib/archivematica.rb:88]: I might be able to do the `"status"` filtering as well at the API level, will look into that.

**Comment by ssciolla** [comment]
Curious to hear what people think of the tests. I'll create a ticket for converting to RSpec, or `minitest/spec`, and looking at [`webmock`](https://github.com/bblimke/webmock).

File: lib/archivematica.rb (hunk 2 above)
Comment (ssciolla) [reply lib/archivematica.rb]: Yep, can filter that using the REST API, though it's not well documented. See my change here: https://github.com/ssciolla/bag-courier/pull/2/commits/82569733cbc4a94f8b808ebf02bde6e057685205

File: run_dark_blue.rb [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+  LOGGER.info(package)
+  id = File.basename(package.path).gsub("-" + package.uuid, "")
+  LOGGER.info("Object ID/Title: #{id}")
+  # Copy file to local source directory here, using SFTP or shared mount?
+  # Extract metadata if possible?
+  # context = nil # maybe some way to determine what content type?
+  # work = Dispatcher::Work.new(
+  #   id: id,
+  #   creator: "Unknown",
+  #   description: "Unknown",
```
Comment (ssciolla) [inline run_dark_blue.rb:43]: John suggests using "Not available" over "Unknown"

**Review by conorom — APPROVED** [review]
Spent a solid half hour looking at this and I genuinely think it all looks great.
I can't think of any nitpick worth mentioning. Down the road I'd probably ask you to add comments to `lib/archivematica.rb` for those not initiated in Archivematica, its AIP's, its API (haha) and the paginated(?) package listings therein... and so forth. But that can totally wait until other source modules get added here. Assuming that happens ;-)
[edit] Plopping links here as I tidy my tabs:
https://www.archivematica.org/en/docs/archivematica-1.10/user-manual/archival-storage/aip-structure/
https://www.archivematica.org/en/docs/archivematica-1.13/dev-manual/api/api-reference-storage-service/#ss-package

**Pull request merged**